### ISSUE TITLE
[New Resource]: AWS SSM tunnel

### ISSUE THREAD

**Issue by dm3ch** [hashicorp/terraform-provider-aws#40249]
### Description

In some cases you may need to connect terraform provider to for example database that are unavailable from the machine running terraform. AWS have aws ssm port-forwarding as a default proposed method to access resources that are only available via private network.

As maintainer wrote it of comunnity-maintained PostgreSQL provider it would be not effective to make postgres provider to work with cloud specific tunneling solutions for each cloud - https://github.com/cyrilgdn/terraform-provider-postgresql/issues/81#issuecomment-901691119. Cause this functionality can be reused between different providers.

In terraform issue https://github.com/hashicorp/terraform/issues/8367#issuecomment-2138275933  itself it was proposed to create separate ephemeral resource for such tunnel. So I propose to consider adding such ephemeral resource to handle aws ssm tunnel creation and later use it in other providers configuration

### Requested Resource(s) and/or Data Source(s)

* aws_ssn_tunnel - like it was proposed at  https://github.com/hashicorp/terraform/issues/8367#issuecomment-2138275933

### Potential Terraform Configuration

```terraform
# INVALID: This is a hypothetical new concept of "ephemeral resources"
# that we're currently considering as part of a solution to this issue.
ephemeral "aws_ssm_tunnel" "mysql_server" {
  instance_id = aws_instance.example.id
  remote_port = 3306
}

provider "mysql" {
  # The local_endpoint attribute would be something like 127.0.0.1:33453,
  # where 33453 is a randomly-selected TCP port number that serves
  # as the local end of the tunnel.
  endpoint = ephemeral.aws_ssm_tunnel.mysql_server.local_endpoint

  # ...
}
```
```


### References

_No response_

### Would you like to implement a fix?

None

**Comment by sdemjanenko** [comment]
This would be amazing. I wrote a custom provider to do exactly this, but there are some edge cases where a keepalive resource is needed to keep the tunnel alive until resources are modified. Also it doesn't work well when removing resources. See: https://registry.terraform.io/providers/ComplyCo/aws-ssm-tunnels/latest

**Comment by dm3ch** [comment]
@sdemjanenko Thank you I would take a look.

To be honest I was going to write provider myself if AWS provider wouldn't have such feature, but I'm happy that it's not needed, cause someone already done this

**Comment by dm3ch** [comment]
@sdemjanenko After a quick look I see that you are using standard resources and not ephemeral ones, isn't it?

As far as I understand the conversion of resources to ephemeral (with TF 1.10) would allow to bind to lifecycle hooks that would allow it to use during apply, plan, destroy without any issues. If I right understood initial discussion in https://github.com/hashicorp/terraform/issues/8367

**Comment by sdemjanenko** [comment]
@dm3ch yes, i am planning on switching to ephemeral once 1.10 hits GA. I am not certain that the lifecycle will be properly tracked (allowing removal of the keepalive resources). That is something I need to test.

From my cursory understanding of ephemeral resources, I believe it will only allow me to get rid of the UUID trick I do to force the tunnel to get set up in the Apply step (https://github.com/ComplyCo/terraform-provider-aws-ssm-tunnels/blob/main/internal/provider/remote_tunnel_resource.go#L192). But this is something I plan on testing to see how much it can simplify my provider.

**Comment by dm3ch** [comment]
jFYI: Now there's a PR to @sdemjanenko provider with ephemeral resource (https://github.com/ComplyCo/terraform-provider-aws-ssm-tunnels/pull/89) which allowed me during tests without any data resources

**Comment by arnaud-dfns** [comment]
Here is my PR adding ephemeral SSM tunnels https://github.com/dfns/terraform-provider-tunnel/pull/2
Will release a new provider version soon

**Comment by arnaud-dfns** [comment]
@dm3ch, it is now released in `v1.1.0` of the provider. You can use it now, example:
```hcl
terraform {
  required_providers {
    tunnel = {
      source  = "dfns/tunnel"
      version = ">= 1.1.0"
    }
  }
}

ephemeral "tunnel_ssm" "eks" {
  target_host  = "https://eks-cluster.region.eks.amazonaws.com"
  target_port  = 443
  ssm_instance = "i-instanceid"
  ssm_region   = "us-east-1"
}

provider "kubernetes" {
  host = "https://${ephemeral.tunnel_ssm.eks.local_host}:${ephemeral.tunnel_ssm.eks.local_port}"

  tls_server_name = "eks-cluster.region.eks.amazonaws.com"

  client_certificate     = file("~/.kube/client-cert.pem")
  client_key             = file("~/.kube/client-key.pem")
  cluster_ca_certificate = file("~/.kube/cluster-ca-cert.pem")
}
```
